### PR TITLE
fix: Table Lineage - Upstream rendering adjustments

### PR DESCRIPTION
**Pull request by bbsbb**
### Summary of Changes

The current rendering relies on two graphs emerging from a single root.
This PR adds the ability to reflow the direction of relationships described in the upstream_entities response to support this.

Examples below:

<details>
  <summary>Example 1</summary>

```javascript
{
  "downstream_entities": [
    {
      "badges": [],
      "cluster": "demo",
      "database": "hive_table",
      "key": "hive_table://demo.sri_lanka/murphy_taylor_and_rivera",
      "level": 2,
      "name": "murphy_taylor_and_rivera",
      "parent": "hive_table://demo.sri_lanka/howard_harrington_and_mendoza",
      "schema": "sri_lanka",
      "source": "hive_table",
      "usage": 0
    },
    {
      "badges": [],
      "cluster": "demo",
      "database": "hive_table",
      "key": "hive_table://demo.sri_lanka/howard_harrington_and_mendoza",
      "level": 1,
      "name": "howard_harrington_and_mendoza",
      "parent": "hive_table://demo.togo/morris_plc",
      "schema": "sri_lanka",
      "source": "hive_table",
      "usage": 0
    },
    {
      "badges": [],
      "cluster": "demo",
      "database": "hive_table",
      "key": "hive_table://demo.christmas_island/oliver_davis_and_reid",
      "level": 1,
      "name": "oliver_davis_and_reid",
      "parent": "hive_table://demo.togo/morris_plc",
      "schema": "christmas_island",
      "source": "hive_table",
      "usage": 0
    },
    {
      "badges": [],
      "cluster": "demo",
      "database": "hive_table",
      "key": "hive_table://demo.sri_lanka/murphy_taylor_and_rivera",
      "level": 2,
      "name": "murphy_taylor_and_rivera",
      "parent": "hive_table://demo.christmas_island/oliver_davis_and_reid",
      "schema": "sri_lanka",
      "source": "hive_table",
      "usage": 0
    }
  ],
  "upstream_entities": [
    {
      "badges": [],
      "cluster": "demo",
      "database": "hive_table",
      "key": "hive_table://demo.sri_lanka/alvarez_and_sons",
      "level": 2,
      "name": "alvarez_and_sons",
      "parent": "",
      "schema": "sri_lanka",
      "source": "hive_table",
      "usage": 0
    },
    {
      "badges": [],
      "cluster": "demo",
      "database": "hive_table",
      "key": "hive_table://demo.austria/owen_lester",
      "level": 1,
      "name": "owen_lester",
      "parent": "hive_table://demo.sri_lanka/alvarez_and_sons",
      "schema": "austria",
      "source": "hive_table",
      "usage": 0
    },
    {
      "badges": [],
      "cluster": "demo",
      "database": "hive_table",
      "key": "hive_table://demo.austria/peterson_plc",
      "level": 1,
      "name": "peterson_plc",
      "parent": "",
      "schema": "austria",
      "source": "hive_table",
      "usage": 0
    }
  ]
}
```
Expected:
lineage_desired_1

Actual:

lineage_actual_1
</details>

<details>
  <summary>Example 2</summary>

```javascript
{
  "downstream_entities": [
    {
      "badges": [],
      "cluster": "demo",
      "database": "hive_table",
      "key": "hive_table://demo.sri_lanka/murphy_taylor_and_rivera",
      "level": 1,
      "name": "murphy_taylor_and_rivera",
      "parent": "hive_table://demo.sri_lanka/howard_harrington_and_mendoza",
      "schema": "sri_lanka",
      "source": "hive_table",
      "usage": 0
    }
  ],
  "upstream_entities": [
    {
      "badges": [],
      "cluster": "demo",
      "database": "hive_table",
      "key": "hive_table://demo.sri_lanka/alvarez_and_sons",
      "level": 3,
      "name": "alvarez_and_sons",
      "parent": "",
      "schema": "sri_lanka",
      "source": "hive_table",
      "usage": 0
    },
    {
      "badges": [],
      "cluster": "demo",
      "database": "hive_table",
      "key": "hive_table://demo.togo/morris_plc",
      "level": 1,
      "name": "morris_plc",
      "parent": "hive_table://demo.austria/peterson_plc",
      "schema": "togo",
      "source": "hive_table",
      "usage": 0
    },
     {                                                                                                                                                                                      
       "badges": [],                                                                                                                                                                          
       "cluster": "demo",                                                                                                                                                                     
       "database": "hive_table",                                                                                                                                                              
       "key": "hive_table://demo.togo/morris_plc",                                                                                                                                            
       "level": 1,                                                                                                                                                                            
       "name": "morris_plc",                                                                                                                                                                  
       "parent": "hive_table://demo.austria/owen_lester",                                                                                                                                     
       "schema": "togo",                                                                                                                                                                      
       "source": "hive_table",                                                                                                                                                                
       "usage": 0                                                                                                                                                                             
     }, 
    {
      "badges": [],
      "cluster": "demo",
      "database": "hive_table",
      "key": "hive_table://demo.austria/owen_lester",
      "level": 2,
      "name": "owen_lester",
      "parent": "hive_table://demo.sri_lanka/alvarez_and_sons",
      "schema": "austria",
      "source": "hive_table",
      "usage": 0
    },
    {
      "badges": [],
      "cluster": "demo",
      "database": "hive_table",
      "key": "hive_table://demo.austria/peterson_plc",
      "level": 2,
      "name": "peterson_plc",
      "parent": "",
      "schema": "austria",
      "source": "hive_table",
      "usage": 0
    }
  ]
}
```

Expected:
lineage_desired_2

Actual:

lineage_actual_2

</details>

### Tests

Added tests to confirm correct upstream lineage transformation.
Added lineage sample to global state Fixture and support for actually testing the successful graph rendering.

### CheckList
Make sure you have checked **all** steps below to ensure a timely review.
- [ ] PR title addresses the issue accurately and concisely. Example: "Updates the version of Flask to v1.0.2"
    - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
- [ ] PR includes a summary of changes.
- [ ] PR adds unit tests, updates existing unit tests, __OR__ documents why no test additions or modifications are needed.
- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
    - All the public functions and the classes in the PR contain docstrings that explain what it does

### CODE REVIEW

**Comment by bbsbb** [comment]
@mgorsk1 - Changes we were discussing - implementation of Example 2 currently relies on API change that matches upstream/downstream response structure.

**Comment by mgorsk1** [comment]
I've aligned to this in https://github.com/amundsen-io/amundsen/pull/1293 but will wait with merging until rendering issues we've discussed are ironed out. Thanks!

**Comment by bbsbb** [comment]
@mgorsk1  - Ready!
@Golodhros  - Would really appreciate a review as you are the person most familiar with this code ;o)

File: frontend/amundsen_application/static/js/components/Lineage/Graph/chart.spec.ts [hunk 1]
```diff
@@ -48,6 +51,18 @@ const nodeWithParent = {
 };
 
 describe('D3 Lineage chart', () => {
+  describe('Generate of shadow node ID should be consistent', () => {
```
Comment (Golodhros) [inline frontend/amundsen_application/static/js/components/Lineage/Graph/chart.spec.ts:54]: Nit: The test suite here seems a bit disorganized right now. You could fix it in different ways, but what I would do is something like:
```
 describe('D3 Lineage chart`
     describe('render', ...
        describe('when data is of regular shape'
        describe('when data is X
        describe('when data is Y
     describe('lifetime', ...
        describe('when user clicks in Z
        describe('when updated with W
        ```

File: frontend/amundsen_application/static/js/components/Lineage/Graph/chart.ts [hunk 2]
```diff
@@ -100,35 +149,61 @@ export const compactLineage = (
...
         n.data.data._parents.includes(p.data.data.key)
       );
 
-      const newX = parents.reduce((sum, p) => sum + p.x, 0) / parents.length;
-
-      acc = acc.concat(
-        parents.map((p) => {
-          if (!p.children && !p._children) {
+      // Determine layout position of the node based on number of parents.
+      n.x = parents.reduce((sum: number, p) => sum + p.x, 0) / parents.length;
```
Comment (Golodhros) [inline frontend/amundsen_application/static/js/components/Lineage/Graph/chart.ts:166]: I would extract a addXValues or similar function

File: frontend/amundsen_application/static/js/components/Lineage/Graph/chart.ts [hunk 3]
```diff
@@ -172,18 +247,10 @@ export const buildEdges = (g, targetNode, nodes) => {
...
     .selectAll('g.graph-node')
     // eslint-disable-next-line no-return-assign
-    .data(nodes, (d) => {
-      if (!d.id) {
-        uniqueIdSequence += 1;
-        d.id = uniqueIdSequence;
-      }
-      return d.id;
-    });
+    .data(nodes, (d) => d.id);
```
Comment (Golodhros) [inline frontend/amundsen_application/static/js/components/Lineage/Graph/chart.ts:253]: nit: ({id} => id)

or 
const getId = ({id}) => id;

File: frontend/amundsen_application/static/js/components/Lineage/Graph/chart.ts [hunk 4]
```diff
@@ -337,64 +403,76 @@ const lc = (): LineageChart => {
...
         (LINEAGE_SCENE_MARGIN.left + LINEAGE_SCENE_MARGIN.right)) /
         2,
     ]);
 
-    const upstreamRoot =
+    const upstreamTree =
       lineage.upstream_entities.length > 0
-        ? hierarchy(stratify(lineage.upstream_entities), (d) => d.children)
+        ? treemap(
+            hierarchy(stratify(lineage.upstream_entities), (d) => d.children)
```
Comment (Golodhros) [inline frontend/amundsen_application/static/js/components/Lineage/Graph/chart.ts:417]: nit: getChildren

File: frontend/amundsen_application/static/js/components/Lineage/Graph/chart.ts [hunk 5]
```diff
@@ -337,64 +403,76 @@ const lc = (): LineageChart => {
...
+    // jsdom does not support SVGAnimation, to test rendering
+    // disable if running inside jest.
+    if (
+      typeof navigator.userAgent !== 'string' ||
+      !navigator.userAgent.includes('jsdom')
+    ) {
+      const zoom = d3Zoom().on('zoom', (e) => {
+        const { transform } = e;
+        // By default make sure to place the graph in center
+        if (!e.sourceEvent) transform.x = dimensions.width / 2;
```
Comment (Golodhros) [inline frontend/amundsen_application/static/js/components/Lineage/Graph/chart.ts:436]: Please use {} with all branching operators

File: frontend/amundsen_application/static/js/components/Lineage/Graph/chart.ts [hunk 6]
```diff
@@ -337,64 +403,76 @@ const lc = (): LineageChart => {
...
     function renderRelativeTo(targetPosition: TreeLineageNode) {
       // Transpose upstream over Y axis
+      const upstreamNodes =
+        upstreamTree?.descendants().map((c) => {
+          // Reorient only if not already done.
+          if (c.y > 0) {
+            c.y *= -1;
+          }
+          return c;
+        }) || [];
```
Comment (Golodhros) [inline frontend/amundsen_application/static/js/components/Lineage/Graph/chart.ts:453]: I would extract a function from this, like 'reorientNode' or similar

File: frontend/amundsen_application/static/js/components/Lineage/Graph/chart.ts [hunk 7]
```diff
@@ -337,64 +403,76 @@ const lc = (): LineageChart => {
...
-        (n: TreeLineageNode): TreeLineageNode => {
+        (n: TreeLineageNode, idx: number): TreeLineageNode => {
           n.x0 = n.x;
           n.y0 = n.y;
+          if (!n.id) {
+            // @ts-ignore
+            n.id = generateNodeId(idx, 0);
+          }
           return n;
         }
```
Comment (Golodhros) [inline frontend/amundsen_application/static/js/components/Lineage/Graph/chart.ts:466]: I would extract a function here too

File: frontend/amundsen_application/static/js/components/Lineage/Graph/chart.ts [hunk 8]
```diff
@@ -337,64 +403,76 @@ const lc = (): LineageChart => {
...
-      const dagNodes = decompactLineage(nodes as TreeLineageNode[]);
-      buildNodes(g, targetPosition, dagNodes, toggler(renderRelativeTo));
+      const nodesToRender = decompactLineage(nodes as TreeLineageNode[]);
+
+      buildNodes(g, targetPosition, nodesToRender, toggler(renderRelativeTo));
       buildEdges(
         g,
         targetPosition,
-        dagNodes.filter((n) => n.depth !== 0)
+        nodesToRender.filter((n) => n.depth !== 0)
```
Comment (Golodhros) [inline frontend/amundsen_application/static/js/components/Lineage/Graph/chart.ts:475]: hasDepth ?

**Review by Golodhros — COMMENTED** [review]
Looks good in general, just some minor comments

File: frontend/amundsen_application/static/js/components/Lineage/Graph/chart.ts [hunk 9]
```diff
@@ -73,6 +82,46 @@ const renderLabels = (
...
+      ...acc,
+      [i.key]: i,
+    }),
+    {}
+  );
+
+  return items.reduce((acc, item) => {
+    if (item.level === 0) {
+      // Add the root node
+      acc.push(item);
```
Comment (Golodhros) [inline frontend/amundsen_application/static/js/components/Lineage/Graph/chart.ts:106]: In general, I would suggest using ... instead of push most of the times. This might be problematic on some cases, but as a general rule, not mutating objects is a good rule of thumb.

Comment (bbsbb) [reply frontend/amundsen_application/static/js/components/Lineage/Graph/chart.ts]: Hey, fully agree. Unfortunately there are a fair bit of in place mutations in this file, but very early I realized that d3 keeps internal references and actually copying/spreading an object quickly makes the behaviour unexpected ;o(
For this function though, it was completely ok to rework without mutations and have done so.

**Comment by bbsbb** [comment]
@Golodhros  - I believe I have addressed the feedback. Also resorted to using argument destructuring in some places where the lambdas were just property accessors.
Have also confirmed functionality with Atlas/Neo4j users. Ultimately, maybe we should add a more comprehensive test set with edge cases for the actual rendering of the lineage graphs, but I don't believe it's object to this PR.

**Review by Golodhros — APPROVED** [review]
LGTM

Thanks!
